### PR TITLE
[WIP] Support autoscaling in the web console

### PR DESCRIPTION
**Pull request by spadgett**
Not ready for code review.

Work in progress PR that adds autoscaling to the web console (reopened from #6552).

https://trello.com/c/SRc3zRmO

TODO:

- [x] Edit or create autoscalers after creation
- [x] Delete autoscalers
- [x] Show autoscaling on overview
- [x] Show autoscaling warnings on overview
- [x] Warn when no CPU request for autoscaling in web console create form
- [x] Warn when no CPU request on standalone autoscaling form
- [x] Handle default values
- [x] Warn when more than one HPA for the same scalable
- [x] Handle autoscaling with limit/request overrides
- [x] Support autoscaling RCs with no DC
- [x] Don't show DC HPAs for on RC page for old deployments
- [x] Warn when autoscaling is set, but CPU request is not
- [x] Warn when metrics not configured
- [x] Warn when two HPAs scale the same resource
- [x] Warn when an RC is scaled by both an HPA and a DC
- [x] Use alert styles for inline warnings
- [x] Delete HPA with DC/RC
- [ ] Spec tests for ApplicationGenerator changes
- [x] Update online-extensions.js for new doc link

/cc @jwforres @DirectXMan12 @ajacobs21e

### CODE REVIEW

**Comment by spadgett** [comment]
Replicas section in create flow, although I'm wondering if this isn't better set later on.

openshift_web_console

Deployment Config:

autoscaling

**Comment by spadgett** [comment]
@DirectXMan12 should we disable manual scaling of deployment configs and replication controllers that have an HPA?

**Comment by spadgett** [comment]
How it looks on the overview currently. Open to suggestions to improve the UI. 

openshift_web_console

I feel like we should show usage vs target visually somehow, here or on the browse pages. A donut might make sense, but we've already used that for the pods. Progress bar?

https://www.patternfly.org/widgets/#progress

**Comment by spadgett** [comment]
Added a link to remove the autoscaler:

openshift_web_console

**Comment by spadgett** [comment]
Working on adding autoscalers to an existing DC.

I'd like to make min and max replicas separate form groups for a few reasons.

1. Min is optional and max is required.
1. We can show errors for one of the fields specifically when the value is invalid.
1. We can have consistent field styles that span the page width. It looks really strange when one field is 100% but another field group only takes part of the page width. For instance, having "Name" stretch across and then the other fields use a small width on this page does not look good.

edit-autoscaler

**Comment by DirectXMan12** [comment]
> @DirectXMan12 should we disable manual scaling of deployment configs and replication controllers that have an HPA?

Yes, probably -- you most likely don't want to be fighting with the autoscaler.

**Comment by DirectXMan12** [comment]
I feel like we should probably show the "CPU request" somewhere (and probably warn if request is unset), since the autoscaler works based on CPU request being set.

**Comment by kargakis** [comment]
> @DirectXMan12 should we disable manual scaling of deployment configs and replication controllers that have an HPA?

We should consider this for oc, too cc: @ironcladlou @smarterclayton

**Comment by spadgett** [comment]
@DirectXMan12 

> I feel like we should probably show the "CPU request" somewhere (and probably warn if request is unset), since the autoscaler works based on CPU request being set.

We show CPU request in the pod template. We could also show it beside the autoscaler, although I'm not sure how that would look with different containers with different requests.

Have a TODO to warn about request not being set.

Is there ever a situation where you would want multiple HPA for the same resource? I plan to flag that as a warning also.

**Comment by DirectXMan12** [comment]
> Have a TODO to warn about request not being set.

:+1:

> Is there ever a situation where you would want multiple HPA for the same resource? I plan to flag that as a warning also.

You mean for the same scalable (RC, DC, etc)?  I don't think so, because they'd end up fighting each other.

**Comment by jwforres** [comment]
On Tue, Mar 1, 2016 at 2:32 PM, Solly Ross <notifications@github.com> wrote:

> Have a TODO to warn about request not being set.
>
> [image: :+1:]
>
> Is there ever a situation where you would want multiple HPA for the same
> resource? I plan to flag that as a warning also.
>
> You mean for the same scalable (RC, DC, etc)? I don't think so, because
> they'd end up fighting each other.
>

Is this something we should warn about in `oc status` as well? @deads2k

**Comment by DirectXMan12** [comment]
@jwforres yeah, I think #7722 is about that

**Comment by deads2k** [comment]
> Yes, probably -- you most likely don't want to be fighting with the autoscaler.

I would expect the autoscaler to have an acceptable range and manual scale requests can take you to the high or low end of that range without being overridden.  Is that not the case for HPA?

**Comment by deads2k** [comment]
> Is this something we should warn about in `oc status` as well? @deads2k

Yes, it's similar to what we do with RCs today. @kargakis you going to roll that into your HPA status issue?

**Comment by kargakis** [comment]
@deads2k done

**Comment by spadgett** [comment]
openshift_web_console

**Comment by jwforres** [comment]
Explain why we are warning them? Something like:
"More than one autoscaler is scaling this resource.  This is not
recommended because the autoscalers may compete with each other."

**Comment by jwforres** [comment]
Would be good to check what warning text was being added to oc status.

**Comment by spadgett** [comment]
> Would be good to check what warning text was being added to oc status.

Looks like it hasn't been added yet (#7722).

**Comment by spadgett** [comment]
Experimenting with progress bars. I think I like this better than the plain text, but probably needs some clarification (66% used of what?). Not sure about placement either.

openshift_web_console

**Comment by spadgett** [comment]
@DirectXMan12 is the default target CPU utilization always 80% if not specified?

**Comment by liggitt** [comment]
it is defaulted to 80% in extensions/v1beta1, and implicit in autoscaling/v1

**Comment by spadgett** [comment]
openshift_web_console

**Comment by ajacobs21e** [comment]
@spadgett I like the progress bar two screenshots ago that's taller and has text inside of it, although I'd recommend the text "using xx% of target". 

Here's a quick sketch of another way to show this that isn't using the progress bar since this isn't really showing progress.
screen shot 2016-03-04 at 1 33 56 am
The top two show how far you are from the target. I think that concept makes more sense than percentage of the target. I also liked your idea of showing target in the middle and distance below or above as long as that visual is intuitive.

The bottom row shows the amount above or below the target with the line reiterating if you're above or below. The color can change to show when you're getting too far from the target.

**Comment by DirectXMan12** [comment]
Note that in the case of the implicit value, you may not be able to get a target percentage (if you use `autoscaling/v1` to fetch the value, as well as decisions made as to the future of matters discussed in kubernetes/kubernetes#22431).  Any UI element should be prepared to just be able to show percentage of requested CPU used, or nothing at all.

With regards to "using xx% of target", would that mean you were recalculating current CPU used with regards to the target for the HPA, or would it just be using the value from the HPA status (which is actually "percentage of requested CPU")?  In either case, IMO the wording on that is a bit confusing.

**Comment by spadgett** [comment]
> With regards to "using xx% of target", would that mean you were recalculating current CPU used with regards to the target for the HPA

@DirectXMan12 It's this, but I agree it's confusing. Right now we're thinking of just saying that this deployment config autoscales from (min) to (max) replicas and only show usage details on the browse deployment config page.

**Comment by liggitt** [comment]
you should still be able to pull the cpu percent *used* out of the status, right? you just may not know the target from the spec if unspecified

**Comment by spadgett** [comment]
Yes. The problem is conveying what that number means for autoscaling succinctly in the overview.

**Comment by DirectXMan12** [comment]
@liggitt yeah, you can always pull the current usage (as percent of requested CPU), you just might not be able to get the the HPA's target percentage (which is implicitly 80% if not specified ATM, but could change)

**Comment by spadgett** [comment]
CPU request warning:

openshift_web_console

**Comment by jwforres** [comment]
i feel like if we are going to inline the warning then maybe it should have
warning alert style?

also looks like a typo "does not have any"

**Comment by spadgett** [comment]
We've done these inline warnings in a few places. I don't have a strong opinion on the style, but I prefer it to a tooltip. I like that you can add corrective actions inline, and it works well on mobile.

Will fix the typo, thanks.

**Comment by DirectXMan12** [comment]
@spadgett forgive me if this is the terminology that the web console uses, but shouldn't the link say "Set resource requests"?

**Comment by spadgett** [comment]
The web console currently has a menu item "Set Resource Limits" that lets you set both requests and limits. I could update the alert link text here to something more specific...  Maybe "Set resource requests and limits"

**Comment by DirectXMan12** [comment]
@spadgett that would probably be less confusing, since HPA is governed specifically by the requests and not by the limits, and it's probably better not to overload terminology when possible.

**Comment by spadgett** [comment]
screen shot 2016-03-10 at 1 34 46 pm

**Comment by DirectXMan12** [comment]
That looks good to me!

File: assets/app/scripts/constants.js [hunk 1]
```diff
@@ -15,6 +15,7 @@ window.OPENSHIFT_CONSTANTS = {
     "route-types":             "https://docs.openshift.org/latest/architecture/core_concepts/routes.html#route-types",
     "persistent_volumes":      "https://docs.openshift.org/latest/dev_guide/persistent_volumes.html",
     "compute_resources":       "https://docs.openshift.org/latest/dev_guide/compute_resources.html",
+    "pod_autoscaling":         "https://docs.openshift.org/latest/dev_guide/pod_autoscaling.html",
```
Comment (jwforres) [inline assets/app/scripts/constants.js]: you'll want to update extensions/examples/online-extensions.js, we should probably add a comment above this section that says to do that.

Comment (spadgett) [inline assets/app/scripts/constants.js]: Thanks, updated.

**Comment by jwforres** [comment]
just thought of something, because we don't surface HPAs as their own resource, what happens when you delete a DC (or RC) that has an HPA associated with it?

**Comment by spadgett** [comment]
> just thought of something, because we don't surface HPAs as their own resource, what happens when you delete a DC (or RC) that has an HPA associated with it?

Currently, the HPA is not deleted. You can only see it from the CLI.

**Comment by liggitt** [comment]
And the controller will keep trying to auto scale based on it :-(

**Comment by kargakis** [comment]
> just thought of something, because we don't surface HPAs as their own resource, what happens when you delete a DC (or RC) that has an HPA associated with it?

I asked about reaping those as well upstream https://github.com/kubernetes/kubernetes/pull/22466#issuecomment-192228463. @liggitt @deads2k how does that sound? Makes sense to me...

**Comment by derekwaynecarr** [comment]
@kargakis - Traditionally, we only cascade delete references that are in one-direction, i.e. an RC will delete pods it manages.  In this case, I think the reference is in the other direction, i.e. the HPA references the DC (or RC), so it feels awkward to set a precedent to delete things that reference a deleted object.

**Comment by smarterclayton** [comment]
I don't think we're ready for that.  Sounds like a 1.3 topic.

**Comment by spadgett** [comment]
Replaced by https://github.com/jwforres/origin/pull/24 against the web console 1.3 branch.

**Pull request closed without merging**